File: backend/app/schemas/subscription.py
"""
Subscription & Billing Pydantic Schemas
"""
from pydantic import BaseModel, EmailStr, Field, field_validator
from typing import Optional, List
from datetime import datetime
from enum import Enum


# ===== ENUMS =====

class PlanType(str, Enum):
    FREE = "free"
    INDIVIDUAL = "individual"
    ORGANIZATION = "organization"
    PAY_AS_YOU_GO = "pay_as_you_go"


class SubscriptionStatus(str, Enum):
    ACTIVE = "active"
    CANCELLED = "cancelled"
    EXPIRED = "expired"
    TRIAL = "trial"
    SUSPENDED = "suspended"


# ===== ORGANIZATION SCHEMAS =====

class OrganizationBase(BaseModel):
    name: str = Field(..., min_length=3, max_length=200)
    slug: str = Field(..., min_length=3, max_length=100)
    description: Optional[str] = None
    billing_email: Optional[EmailStr] = None


class OrganizationCreate(OrganizationBase):
    max_members: int = Field(default=10, ge=1, le=1000)


class OrganizationUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=3, max_length=200)
    description: Optional[str] = None
    billing_email: Optional[EmailStr] = None
    max_members: Optional[int] = Field(None, ge=1, le=1000)
    is_active: Optional[bool] = None


class OrganizationResponse(OrganizationBase):
    id: int
    owner_id: int
    max_members: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ===== ORGANIZATION MEMBER SCHEMAS =====

class OrganizationMemberAdd(BaseModel):
    user_id: int
    role: str = Field(default="member", pattern="^(owner|admin|member)$")


class OrganizationMemberResponse(BaseModel):
    id: int
    organization_id: int
    user_id: int
    role: str
    is_active: bool
    invited_at: datetime
    joined_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ===== PRICING PLAN SCHEMAS =====

class PricingPlanBase(BaseModel):
    plan_type: PlanType
    name: str
    description: Optional[str] = None
    monthly_price: float = Field(ge=0)
    annual_price: Optional[float] = Field(None, ge=0)


class PricingPlanCreate(PricingPlanBase):
    premium_requests_limit: Optional[int] = None  # AI requests/month
    monthly_spending_limit: Optional[float] = None  # AI credits
    features: Optional[str] = None  # JSON string
    trial_days: int = Field(default=14, ge=0, le=365)


class PricingPlanUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    monthly_price: Optional[float] = Field(None, ge=0)
    annual_price: Optional[float] = Field(None, ge=0)
    premium_requests_limit: Optional[int] = None  # AI requests
    monthly_spending_limit: Optional[float] = None  # AI credits
    features: Optional[str] = None
    is_active: Optional[bool] = None
    is_public: Optional[bool] = None
    trial_days: Optional[int] = Field(None, ge=0, le=365)


class PricingPlanResponse(PricingPlanBase):
    id: int
    premium_requests_limit: Optional[int]  # AI requests/month
    monthly_spending_limit: Optional[float]  # AI credits
    features: Optional[str]
    is_active: bool
    is_public: bool
    trial_days: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ===== SUBSCRIPTION SCHEMAS =====

class SubscriptionBase(BaseModel):
    plan_type: PlanType
    monthly_limit_usd: Optional[float] = Field(None, ge=0)


class SubscriptionCreate(SubscriptionBase):
    organization_id: Optional[int] = None


class SubscriptionUpdate(BaseModel):
    plan_type: Optional[PlanType] = None
    monthly_limit_usd: Optional[float] = Field(None, ge=0)
    cancel_at_period_end: Optional[bool] = None


class SubscriptionResponse(BaseModel):
    id: int
    user_id: Optional[int]
    organization_id: Optional[int]
    plan_type: PlanType
    status: SubscriptionStatus
    monthly_price: float
    monthly_limit_usd: Optional[float]
    # Map từ model fields (premium_requests_limit)
    premium_requests_limit: Optional[int] = None
    premium_requests_used: Optional[int] = None
    current_period_start: Optional[datetime]
    current_period_end: Optional[datetime]
    trial_start: Optional[datetime]
    trial_end: Optional[datetime]
    cancel_at_period_end: bool
    cancelled_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ===== USAGE SCHEMAS =====

class UsageSummary(BaseModel):
    """Current period usage summary"""
    period_start: datetime
    period_end: datetime
    total_requests: int
    total_tokens: int
    total_cost: float
    
    # Breakdown by provider
    gemini_cost: float
    claude_cost: float
    adobe_cost: float
    
    # Limits
    monthly_limit: Optional[float]
    remaining_budget: Optional[float]
    usage_percentage: Optional[float]


class DailyUsage(BaseModel):
    """Daily usage statistics"""
    date: str  # YYYY-MM-DD
    requests: int
    tokens: int
    cost: float


class ProviderUsage(BaseModel):
    """Usage by provider"""
    provider: str
    requests: int
    tokens: int
    cost: float
    percentage: float


class UsageStatsResponse(BaseModel):
    """Detailed usage statistics"""
    summary: UsageSummary
    daily_usage: List[DailyUsage]
    provider_breakdown: List[ProviderUsage]
    top_operations: List[dict]  # Top 5 operations by cost


# ===== BILLING SCHEMAS =====

class BillingHistoryResponse(BaseModel):
    id: int
    subscription_id: int
    billing_month: str
    period_start: datetime
    period_end: datetime
    total_requests: int
    total_tokens: int
    gemini_cost: float
    claude_cost: float
    adobe_cost: float
    total_cost: float
    subscription_fee: float
    total_amount: float
    status: str
    paid_at: Optional[datetime]
    invoice_number: Optional[str]
    invoice_url: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True


class BillingListResponse(BaseModel):
    """List of billing records with pagination"""
    total: int
    page: int
    page_size: int
    total_pages: int
    records: List[BillingHistoryResponse]


# ===== DASHBOARD SCHEMAS =====

class UserDashboardResponse(BaseModel):
    """User dashboard overview"""
    user_id: int
    username: str
    email: str
    
    # Subscription info
    subscription: SubscriptionResponse
    
    # Current period usage
    current_usage: UsageSummary
    
    # Alerts
    alerts: List[dict]  # Budget warnings, limit warnings, etc.


class OrganizationDashboardResponse(BaseModel):
    """Organization dashboard overview"""
    organization: OrganizationResponse
    subscription: SubscriptionResponse
    current_usage: UsageSummary
    member_count: int
    top_users: List[dict]  # Top users by usage
    alerts: List[dict]

File: frontend/src/services/subscription.ts
/**
 * Subscription & Billing API Service
 */
import api from './api';

// Types
export interface PlanType {
  FREE: 'free';
  INDIVIDUAL: 'individual';
  ORGANIZATION: 'organization';
  PAY_AS_YOU_GO: 'pay_as_you_go';
}

export interface SubscriptionStatus {
  ACTIVE: 'active';
  CANCELLED: 'cancelled';
  EXPIRED: 'expired';
  TRIAL: 'trial';
  SUSPENDED: 'suspended';
}

export interface PricingPlan {
  id: number;
  plan_type: string;
  name: string;
  description: string | null;
  monthly_price: number;
  annual_price: number | null;
  premium_requests_limit: number | null;  // AI requests limit
  monthly_spending_limit: number | null;  // AI credits
  features: string | null;
  is_active: boolean;
  is_public: boolean;
  trial_days: number;
  created_at: string;
  updated_at: string;
}

export interface Subscription {
  id: number;
  user_id: number | null;
  organization_id: number | null;
  plan_type: string;
  status: string;
  monthly_price: number;
  monthly_limit_usd: number | null;
  premium_requests_used: number;  // Used this period
  premium_requests_limit: number | null;  // From pricing plan
  current_period_start: string | null;
  current_period_end: string | null;
  trial_start: string | null;
  trial_end: string | null;
  cancel_at_period_end: boolean;
  cancelled_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface UsageSummary {
  period_start: string;
  period_end: string;
  total_requests: number;
  total_tokens: number;
  total_cost: number;
  gemini_cost: number;
  claude_cost: number;
  adobe_cost: number;
  monthly_limit: number | null;
  remaining_budget: number | null;
  usage_percentage: number | null;
}

export interface DailyUsage {
  date: string;
  requests: number;
  tokens: number;
  cost: number;
}

export interface ProviderUsage {
  provider: string;
  requests: number;
  tokens: number;
  cost: number;
  percentage: number;
}

export interface UsageStats {
  summary: UsageSummary;
  daily_usage: DailyUsage[];
  provider_breakdown: ProviderUsage[];
  top_operations: Array<{
    operation: string;
    count: number;
    cost: number;
  }>;
}

export interface BillingRecord {
  id: number;
  subscription_id: number;
  billing_month: string;
  period_start: string;
  period_end: string;
  total_requests: number;
  total_tokens: number;
  gemini_cost: number;
  claude_cost: number;
  adobe_cost: number;
  total_cost: number;
  subscription_fee: number;
  total_amount: number;
  status: string;
  paid_at: string | null;
  invoice_number: string | null;
  invoice_url: string | null;
  created_at: string;
}

export interface BillingListResponse {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  records: BillingRecord[];
}

export interface Organization {
  id: number;
  name: string;
  slug: string;
  description: string | null;
  owner_id: number;
  max_members: number;
  is_active: boolean;
  billing_email: string | null;
  created_at: string;
  updated_at: string;
}

// API Service
export const subscriptionService = {
  // Pricing Plans
  getPricingPlans: async (): Promise<PricingPlan[]> => {
    const response = await api.get('/subscription/pricing-plans');
    return response.data;
  },

  getPricingPlan: async (planType: string): Promise<PricingPlan> => {
    const response = await api.get(`/subscription/pricing-plans/${planType}`);
    return response.data;
  },

  // User Subscription
  getMySubscription: async (): Promise<Subscription> => {
    const response = await api.get('/subscription/my-subscription');
    return response.data;
  },

  createSubscription: async (data: {
    plan_type: string;
    organization_id?: number;
    monthly_limit_usd?: number;
  }): Promise<Subscription> => {
    const response = await api.post('/subscription/subscribe', data);
    return response.data;
  },

  updateSubscription: async (data: {
    plan_type?: string;
    monthly_limit_usd?: number;
    cancel_at_period_end?: boolean;
  }): Promise<Subscription> => {
    const response = await api.put('/subscription/my-subscription', data);
    return response.data;
  },

  cancelSubscription: async (): Promise<{ message: string; cancellation_date: string }> => {
    const response = await api.delete('/subscription/my-subscription');
    return response.data;
  },

  // Usage
  getMyUsage: async (): Promise<UsageSummary> => {
    const response = await api.get('/subscription/my-usage');
    return response.data;
  },

  getMyUsageDetailed: async (days: number = 30): Promise<UsageStats> => {
    const response = await api.get(`/subscription/my-usage/detailed?days=${days}`);
    return response.data;
  },

  // Billing
  getMyBilling: async (page: number = 1, pageSize: number = 12): Promise<BillingListResponse> => {
    const response = await api.get(`/subscription/my-billing?page=${page}&page_size=${pageSize}`);
    return response.data;
  },

  getBillingDetail: async (billingId: number): Promise<BillingRecord> => {
    const response = await api.get(`/subscription/my-billing/${billingId}`);
    return response.data;
  },

  // Organizations
  createOrganization: async (data: {
    name: string;
    slug: string;
    description?: string;
    billing_email?: string;
    max_members?: number;
  }): Promise<Organization> => {
    const response = await api.post('/subscription/organizations', data);
    return response.data;
  },

  getMyOrganizations: async (): Promise<Organization[]> => {
    const response = await api.get('/subscription/organizations/my');
    return response.data;
  },

  getOrganization: async (orgId: number): Promise<Organization> => {
    const response = await api.get(`/subscription/organizations/${orgId}`);
    return response.data;
  },
};

export default subscriptionService;
